code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm

```diff
@@ -0,0 +1,487 @@
+// ***************************************
+// *********** Place Warp Shadow
+// ***************************************
+/datum/action/xeno_action/place_warp_shadow
+	name = "Place Warp Shadow"
+	action_icon_state = "warp_shadow"
+	mechanics_text = "Binds our psychic essence to a spot of our choosing. We can use Hyperposition to swap locations with this essence."
+	plasma_cost = 150
+	cooldown_timer = 60 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_PLACE_WARP_BEACON
+
+/datum/action/xeno_action/place_warp_shadow/action_activate()
+	var/mob/living/carbon/xenomorph/wraith/TI = owner
+
+	TI.visible_message("<span class='xenowarning'>The air starts to violently roil and shimmer around [TI]!</span>", \
+	"<span class='xenodanger'>We begin to imprint our essence upon reality, causing the air about us to roil and shimmer...</span>") //Fluff
+
+	if(!do_after(TI, WRAITH_PLACE_WARP_BEACON_WINDUP, TRUE, TI, BUSY_ICON_BUILD)) //Channel time/wind up
+		TI.visible_message("<span class='xenowarning'>The space around [TI] abruptly stops shifting and wavering.</span>", \
+		"<span class='xenodanger'>We cease binding our essence to this place...</span>")
+		return fail_activate()
+
+	var/turf/T = get_turf(TI)
+	TI.visible_message("<span class='xenonotice'>\A shimmering point suddenly coalesces from the warped space around [T].</span>", \
+	"<span class='xenodanger'>We complete our work, binding our essence to this point.</span>", null, 5) //Fluff
+	var/obj/effect/xenomorph/warp_shadow/WB = new(T) //Create the new warp shadow.
+	playsound(T, 'sound/weapons/emitter.ogg', 25, 1)
+	QDEL_NULL(TI.warp_shadow) //Delete the old warp shadow
+	TI.warp_shadow = WB //Set our new warp shadow
+	RegisterSignal(TI.warp_shadow, COMSIG_PARENT_PREQDELETED, .proc/unset_target) //For var clean up
+	WB.setDir(TI.dir) //Have it imitate our facing
+	WB.pixel_x = TI.pixel_x //Inherit pixel offsets
+	WB.pixel_y = TI.pixel_y //Inherit pixel offsets
+	teleport_debuff_aoe(WB) //SFX
+
+	succeed_activate()
+	add_cooldown()
+
+/datum/action/xeno_action/place_warp_shadow/proc/unset_target()
+	var/mob/living/carbon/xenomorph/X = owner
+	UnregisterSignal(X.warp_shadow, COMSIG_PARENT_PREQDELETED)
+	X.warp_shadow = null
+
+
+// ***************************************
+// *********** Hyperposition
+// ***************************************
+/datum/action/xeno_action/hyperposition
+	name = "Hyperposition"
+	action_icon_state = "hyperposition"
+	mechanics_text = "We teleport back to our warp shadow after a delay. The delay scales with the distance teleported."
+	plasma_cost = 25
+	cooldown_timer = 5 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_HYPERPOSITION
+
+
+/datum/action/xeno_action/hyperposition/proc/can_use_ability()
+
+	var/mob/living/carbon/xenomorph/wraith/TI = owner
+	var/obj/effect/xenomorph/warp_shadow/WB = TI.warp_shadow
+	var/beacon_turf = get_turf(WB)
+	var/wraith_turf = get_turf(TI)
+
```

Comment:
move this until after the WB check or more accurately remove them since they arent needed anymore